File: matlab/tests/unit/test_jitter_load.m
%% test_jitter_load.m - Unit test for jitter analysis
close all; clear; clc; warning("off")

%% Configuration
verbose = 0;
inputDir = fullfile('dataset', 'jitter_sweep');
outputDir = fullfile('test_output', 'jitter_sweep');
if ~exist(outputDir, 'dir'), mkdir(outputDir); end

config_filepath = fullfile(inputDir, 'config.csv');
if ~exist(config_filepath, 'file')
    error('[Config file not found] Please run generate_jitter_sweep_data.m first');
end

config_data = readcell(config_filepath);
Fs_expected = config_data{2, 2};
N_expected = config_data{3, 2};

metadata_filepath = fullfile(inputDir, 'jitter_sweep_metadata.csv');
metadata = readmatrix(metadata_filepath);
Tj_list = metadata(:, 2);

freq_metadata_filepath = fullfile(inputDir, 'frequency_list.csv');
freq_metadata = readmatrix(freq_metadata_filepath);
Fin_list_nominal = freq_metadata(:, 1);

%% Analyze each frequency
for i_freq = 1:length(Fin_list_nominal)
    Fin_nominal = Fin_list_nominal(i_freq);
    fprintf('\n[%s] [%d/%d] [Fin = %d MHz]\n', mfilename, i_freq, length(Fin_list_nominal), round(Fin_nominal/1e6));

    meas_jitter = zeros(length(Tj_list), 1);
    meas_SNDR = zeros(length(Tj_list), 1);
    set_jitter = zeros(length(Tj_list), 1);
    actual_Fin = zeros(length(Tj_list), 1);
    pnoi_array = zeros(length(Tj_list), 1);
    anoi_array = zeros(length(Tj_list), 1);

    for i_tj = 1:length(Tj_list)
        Tj = Tj_list(i_tj);
        set_jitter(i_tj) = Tj;

        filename = sprintf('jitter_sweep_Fin_%dMHz_Tj_idx_%02d.csv', ...
            round(Fin_nominal/1e6), i_tj);
        filepath = fullfile(inputDir, filename);

        if ~exist(filepath, 'file')
            warning('[File not found] %s', filepath);
            continue;
        end

        read_data = readmatrix(filepath);
        N = length(read_data);

        if N ~= N_expected
            warning('[N mismatch] Data has N=%d, config expects N=%d', N, N_expected);
        end

        [data_fit, f_norm, mag, dc, phi] = sineFit(read_data);
        Fin_fit = f_norm * Fs_expected;
        actual_Fin(i_tj) = Fin_fit;

        [emean, erms, phase_code, anoi, pnoi] = errHistSine(read_data, 99, f_norm, 0);
        pnoi_array(i_tj) = pnoi;
        anoi_array(i_tj) = anoi;

        jitter_rms = pnoi / (2 * pi * Fin_fit);
        meas_jitter(i_tj) = jitter_rms;

        [ENoB, SNDR, SFDR, SNR, THD, pwr, NF, h] = specPlot(read_data, ...
            'label', 1, 'harmonic', 0, 'winType', @hann, ...
            'OSR', 1, 'coAvg', 0, "isPlot", 0);
        meas_SNDR(i_tj) = SNDR;
    end

    Fin_actual_mean = mean(actual_Fin);

    figure('Position', [100, 100, 1000, 600], "Visible", verbose);
    yyaxis left
    loglog(set_jitter, set_jitter, 'k--', 'LineWidth', 1.5, "DisplayName", "Set jitter");
    hold on;
    loglog(set_jitter, meas_jitter, 'bo', 'MarkerSize', 8, 'MarkerFaceColor', 'b', "DisplayName", "Calculated jitter");
    ylabel("Jitter (seconds)", "FontSize", 18);
    ylim([min(set_jitter) * 0.5, max(set_jitter) * 2]);

    yyaxis right
    semilogx(set_jitter, meas_SNDR, 's-', 'LineWidth', 2, 'MarkerSize', 8, "DisplayName", "SNDR");
    ylabel("SNDR (dB)", "FontSize", 18);
    ylim([0, 100]);

    xlabel("Set jitter (seconds)", "FontSize", 18);
    title(sprintf("Jitter Analysis (Fin = %.1f MHz)", Fin_actual_mean/1e6), "FontSize", 20);
    legend("Location", "southeast", "FontSize", 16);
    grid on;
    set(gca, "FontSize", 16);

    output_filename = sprintf('jitter_analysis_Fin_%dMHz_matlab.png', round(Fin_nominal/1e6));
    saveFig(outputDir, output_filename, verbose);

    results_filename = sprintf('jitter_results_Fin_%dMHz_matlab.csv', round(Fin_nominal/1e6));
    results_filepath = fullfile(outputDir, results_filename);

    fid = fopen(results_filepath, 'w');
    fprintf(fid, 'Tj_idx,set_jitter_s,set_jitter_fs,meas_jitter_s,meas_jitter_fs,pnoi_rad,anoi,SNDR_dB,actual_Fin_Hz\n');
    for i = 1:length(Tj_list)
        fprintf(fid, '%d,%.12e,%.6f,%.12e,%.6f,%.12e,%.12e,%.2f,%.6e\n', ...
            i, set_jitter(i), set_jitter(i)*1e15, ...
            meas_jitter(i), meas_jitter(i)*1e15, ...
            pnoi_array(i), anoi_array(i), meas_SNDR(i), actual_Fin(i));
    end
    fclose(fid);
end

File: python/tests/unit/test_jitter_load.py
"""test_jitter_load.py - Test jitter analysis with deterministic data

Input: test_data/jitter_sweep/
Output: test_output/jitter_sweep/
"""

import numpy as np
import pandas as pd
import matplotlib
matplotlib.use('Agg')  # Non-interactive backend
import matplotlib.pyplot as plt
from pathlib import Path

from adctoolbox.common import sine_fit
from adctoolbox.aout import spec_plot, err_hist_sine
from tests._utils import save_variable, save_fig

# Get project root directory
project_root = Path(__file__).resolve().parents[3]

def main():
    """Main test function."""
    input_dir = project_root / "dataset" / "jitter_sweep"
    output_dir = project_root / "test_output" / "jitter_sweep"

    if not output_dir.exists():
        output_dir.mkdir(parents=True)

    # Load configuration
    config_filepath = input_dir / 'config.csv'
    if not config_filepath.exists():
        print('[ERROR] Config file not found. Please run generate_jitter_sweep_data first')
        return

    config_data = pd.read_csv(config_filepath)
    Fs_expected = config_data.loc[config_data.iloc[:, 0] == 'Fs', config_data.columns[1]].values[0]
    N_expected = int(config_data.loc[config_data.iloc[:, 0] == 'N', config_data.columns[1]].values[0])

    # Load metadata
    metadata_filepath = input_dir / 'jitter_sweep_metadata.csv'
    metadata = pd.read_csv(metadata_filepath)  # Read with header
    Tj_list = metadata['Tj_seconds'].values  # Use column name

    # Load frequency list
    freq_metadata_filepath = input_dir / 'frequency_list.csv'
    freq_metadata = pd.read_csv(freq_metadata_filepath)
    Fin_list_nominal = freq_metadata.iloc[:, 0].astype(float).values

    print('=== test_jitter_load.py ===')
    print(f'[Input dir] {input_dir}')
    print(f'[Output dir] {output_dir}')
    print(f'[Fs from config] {Fs_expected:.3e} Hz')
    print(f'[N from config] {N_expected}\n')

    # Analyze each frequency
    for i_freq, Fin_nominal in enumerate(Fin_list_nominal):
        print(f'[Analyzing] Nominal Fin = {round(Fin_nominal/1e6)} MHz')

        meas_jitter = np.zeros(len(Tj_list))
        meas_SNDR = np.zeros(len(Tj_list))
        set_jitter = np.zeros(len(Tj_list))
        actual_Fin = np.zeros(len(Tj_list))
        pnoi_array = np.zeros(len(Tj_list))
        anoi_array = np.zeros(len(Tj_list))

        for i_tj, Tj in enumerate(Tj_list):
            set_jitter[i_tj] = Tj

            filename = f'jitter_sweep_Fin_{round(Fin_nominal/1e6)}MHz_Tj_idx_{i_tj+1:02d}.csv'
            filepath = input_dir / filename

            if not filepath.exists():
                print(f'[WARNING] File not found: {filepath}')
                continue

            read_data = np.loadtxt(filepath, delimiter=',')
            N = len(read_data)

            if N != N_expected:
                print(f'[WARNING] N mismatch: Data has N={N}, config expects N={N_expected}')

            # Sine fit
            data_fit, f_norm, mag, dc, phi = sine_fit(read_data)
            Fin_fit = f_norm * Fs_expected
            actual_Fin[i_tj] = Fin_fit

            # Error histogram analysis
            emean, erms, phase_code, anoi, pnoi, err, xx = err_hist_sine(
                read_data, bin=99, fin=f_norm, disp=0
            )
            pnoi_array[i_tj] = pnoi
            anoi_array[i_tj] = anoi

            # Calculate jitter
            jitter_rms = pnoi / (2 * np.pi * Fin_fit)
            meas_jitter[i_tj] = jitter_rms

            # Spectrum analysis
            ENoB, SNDR, SFDR, SNR, THD, pwr, NF, _ = spec_plot(
                read_data,
                label=1,
                harmonic=0,
                osr=1,
                nf_method=0
            )
            meas_SNDR[i_tj] = SNDR
            plt.close()  # Close the spectrum plot

            if (i_tj + 1) % 5 == 0:
                print(f'  [{i_tj+1}/{len(Tj_list)}] [test_jitter_load] [Tj={Tj*1e15:.2f}fs→{jitter_rms*1e15:.2f}fs] [SNDR={SNDR:.2f}dB] from {filename}')

        print()

        Fin_actual_mean = np.mean(actual_Fin)
        print(f'[Extracted Fin from data] = {Fin_actual_mean/1e6:.6f} MHz (nominal was {Fin_nominal/1e6:.6f} MHz)')

        # Plot results
        fig = plt.figure(figsize=(10, 6))
        ax1 = plt.gca()

        # Left y-axis: Jitter
        ax1.loglog(set_jitter, set_jitter, 'k--', linewidth=1.5, label='Set jitter')
        ax1.loglog(set_jitter, meas_jitter, 'bo', markersize=8, markerfacecolor='b', label='Calculated jitter')
        ax1.set_xlabel('Set jitter (seconds)', fontsize=18)
        ax1.set_ylabel('Jitter (seconds)', fontsize=18, color='b')
        ax1.set_ylim([min(set_jitter) * 0.5, max(set_jitter) * 2])
        ax1.tick_params(axis='y', labelcolor='b', labelsize=16)
        ax1.tick_params(axis='x', labelsize=16)
        ax1.grid(True)

        # Right y-axis: SNDR
        ax2 = ax1.twinx()
        ax2.semilogx(set_jitter, meas_SNDR, 's-', linewidth=2, markersize=8, color='r', label='SNDR')
        ax2.set_ylabel('SNDR (dB)', fontsize=18, color='r')
        ax2.set_ylim([0, 100])
        ax2.tick_params(axis='y', labelcolor='r', labelsize=16)

        plt.title(f'Jitter Analysis (Fin = {Fin_actual_mean/1e6:.1f} MHz)', fontsize=20)

        # Combine legends
        lines1, labels1 = ax1.get_legend_handles_labels()
        lines2, labels2 = ax2.get_legend_handles_labels()
        ax1.legend(lines1 + lines2, labels1 + labels2, loc='lower right', fontsize=16)

        # Save plot
        output_filename = f'jitter_analysis_Fin_{round(Fin_nominal/1e6)}MHz_python.png'
        save_fig(output_dir, output_filename)
        print(f'[Saved plot] -> [{output_dir / output_filename}]\n')

        # Save each variable to separate CSV (matching MATLAB format)
        # Create frequency-specific prefix for filenames
        freq_prefix = f'Fin_{round(Fin_nominal/1e6)}MHz'
        save_variable(output_dir, np.arange(1, len(Tj_list) + 1), f'{freq_prefix}_Tj_idx')
        save_variable(output_dir, set_jitter, f'{freq_prefix}_set_jitter_s')
        save_variable(output_dir, set_jitter * 1e15, f'{freq_prefix}_set_jitter_fs')
        save_variable(output_dir, meas_jitter, f'{freq_prefix}_meas_jitter_s')
        save_variable(output_dir, meas_jitter * 1e15, f'{freq_prefix}_meas_jitter_fs')
        save_variable(output_dir, pnoi_array, f'{freq_prefix}_pnoi_rad')
        save_variable(output_dir, anoi_array, f'{freq_prefix}_anoi')
        save_variable(output_dir, meas_SNDR, f'{freq_prefix}_SNDR_dB')
        save_variable(output_dir, actual_Fin, f'{freq_prefix}_actual_Fin_Hz')

    print('[test_jitter_load complete]')


if __name__ == "__main__":
    main()
